data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/CompressionEngine.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.amazon.dataprepper.plugins.source;
+
+import com.amazon.dataprepper.plugins.source.configuration.CompressionOption;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.zip.GZIPInputStream;
+
+public class CompressionEngine {
+    private final CompressionOption compressionOption;
+
+    public CompressionEngine(CompressionOption compressionOption) {
+        this.compressionOption = compressionOption;
+    }
+
+    InputStream createInputStream(String s3Key, InputStream responseInputStream) throws IOException {
+        if (compressionOption.equals(CompressionOption.NONE))
+            return responseInputStream;
+        else if (compressionOption.equals(CompressionOption.GZIP))
+            return new GZIPInputStream(responseInputStream);
+        else {
+            if (s3Key.endsWith(".gz")) {
+                return new GZIPInputStream(responseInputStream);
+            }
+            else {
+                return responseInputStream;
+            }
+        }
```

Comment:
I have a few comments on this code, 1) There is a disconnect with the CompressionOption and the business logic for applying this compressionOption 2) this pattern of leveraging conditionals does not scale well for adding new options. 3) we have to compute the compression engine for every single.

I would recommend we redesign the current class architecture and create an interface for the compression options that requires a `createInputStream`. The desired compression option can be passed in as class parameter to the `S3ObjectWorker`. This will eliminate the redundant computation, provide a tighter coupling for the option and the implementation and make it easier for contributors to add new options.

Reply:
Thanks for the suggestion @cmanning09. I think if we pass the compression implementation as input to `S3ObjectWorker` we can support both `NONE` and `GZIP`. I can't think of a way right now how we could implement `AUTOMATIC` option without knowing the s3 object key.

Reply:
I think the idea is that `CompressionEngine` itself could be an interface. And in the `S3Source` class we create a `CompressionEngine` for each specific type.

So you might have:

```
class AutomaticCompressionEngine implements CompressionEngine {
  InputStream createInputStream(String s3Key, InputStream responseInputStream) throws IOException {
    if (s3Key.endsWith(".gz")) {
        return new GZIPInputStream(responseInputStream);
    }
    else {
        return responseInputStream;
    }
  }
}
```

And we'd have two other interfaces - `GzipCompressionEngine` and `NoneCompressionEngine` that already know what to do.

Reply:
Exactly, thanks for clarifying @dlvenable

Reply:
Thanks David. It makes it easier to add to add new options in future, but we still don't eliminate redundant computation for each s3 object. I'll make these changes and we can think about an approach to eliminate redundant computation in another PR.

Reply:
> Thanks David and Chris. It makes it easier to add to add new options in future, but we still don't eliminate redundant computation for each s3 object. I'll make these changes and we can think about an approach to eliminate redundant computation in another PR.

This does eliminate the redundant computation that I believe @cmanning09 was speaking of. The conditionals to check against `compressionEngine==Compression.XYZ` had to happen for each S3 object. Now they are done only at startup. Either way, this is not likely to have any significant impact on performance.